### PR TITLE
Horizontally Scalable Ruler

### PR DESCRIPTION
**Pull request by jtlisi**
This PR provides a flag that allows the ruler to shard ownership of which rule groups to evaluate by re-using the ring code. When enabled each ruler will act as a Lifecycler and interact with consul to create a ring.

## TO DO
- [x] Add tests for the ruler ring
- [x] Add better instrumentation for monitoring the distributed ruler

### CODE REVIEW

**Comment by khaines** [comment]
Thank you for working on this change! I'll be very happy to test this when it is ready :)

@rsteneteg FYI, since we've discussed this as a scaling contention point in the past

**Comment by tomwilkie** [comment]
Does this also fix #310?

**Comment by bboreham** [comment]
Fixes #1213.

File: cmd/ruler/main.go [hunk 1]
```diff
@@ -59,28 +58,29 @@ func main() {
...
-	util.CheckFatal("initializing ruler", err)
-	defer rlr.Stop()
 
-	rulesAPI, err := ruler.NewRulesAPI(configStoreConfig)
-	util.CheckFatal("initializing rules API", err)
-
-	rulerServer, err := ruler.NewServer(rulerConfig, rlr, rulesAPI)
-	util.CheckFatal("initializing ruler server", err)
-	defer rulerServer.Stop()
+	rulerConfig.LifecyclerConfig = lifecyclerConfig
```
Comment (tomwilkie) [inline cmd/ruler/main.go]: Can we have the rulerConfig recursively register the lifecycler, so we don't have to do this?  This has been error prone in the past.

File: cmd/ruler/main.go [hunk 2]
```diff
@@ -59,28 +58,29 @@ func main() {
 	util.CheckFatal("", err)
 	defer chunkStore.Stop()
 
-	r, err := ring.New(ringConfig)
+	r, err := ring.New(lifecyclerConfig.RingConfig)
```
Comment (tomwilkie) [inline cmd/ruler/main.go]: I'd like us to have two ringConfigs here - one for the ingesters and one for the rulers.  This will allow us to store them in different consuls in the future.  They should register their flags with different prefixes.

File: pkg/ring/lifecycler.go [hunk 3]
```diff
@@ -19,15 +19,15 @@ import (
 
 var (
 	consulHeartbeats = promauto.NewCounter(prometheus.CounterOpts{
-		Name: "cortex_ingester_consul_heartbeats_total",
+		Name: "cortex_member_consul_heartbeats_total",
```
Comment (tomwilkie) [inline pkg/ring/lifecycler.go]: We should add a label to these metrics so they can differentiate between different rings.

File: pkg/ring/lifecycler.go [hunk 4]
```diff
@@ -97,7 +97,7 @@ type Lifecycler struct {
 
 	// These values are initialised at startup, and never change
 	ID   string
-	addr string
+	Addr string
```
Comment (tomwilkie) [inline pkg/ring/lifecycler.go]: Why was this necessary?

File: pkg/ring/ring.go [hunk 5]
```diff
@@ -100,30 +101,34 @@ func New(cfg Config) (*Ring, error) {
 		return nil, err
 	}
 
+	if cfg.Name == "" {
+		cfg.Name = "ingester"
+	}
```
Comment (tomwilkie) [inline pkg/ring/ring.go]: Lets not do this.  Also, I don't think the name is part of the config - I think they're defined by the code.  Name should be a string passed to the constructor, and should not be the empty string.

File: pkg/ring/ring.go [hunk 6]
```diff
@@ -100,30 +101,34 @@ func New(cfg Config) (*Ring, error) {
...
 		done:     make(chan struct{}),
 		ringDesc: &Desc{},
-		ingesterOwnershipDesc: prometheus.NewDesc(
-			"cortex_ring_ingester_ownership_percent",
-			"The percent ownership of the ring by ingester",
-			[]string{"ingester"}, nil,
+		memberOwnershipDesc: prometheus.NewDesc(
+			"cortex_ring_member_ownership_percent",
+			"The percent ownership of the ring by member",
+			[]string{"member", "type"}, nil,
```
Comment (tomwilkie) [inline pkg/ring/ring.go]: s/type/name/?

File: pkg/ruler/lifecycle.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+				time.Sleep(r.SearchPendingFor / pendingSearchIterations)
+				continue
+			} else {
+				level.Warn(util.Logger).Log("msg", "Could not find pending ruler before deadline", "err", err)
+				return nil, err
+			}
+		}
+		return ingester, nil
+	}
+}
```
Comment (tomwilkie) [inline pkg/ruler/lifecycle.go]: Is this actually used?  Does it need to exist?  I think transfer in the rulers case is a noop.

File: pkg/ruler/ring.go [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+	consulCfg.Prefix = "rulers/"
+
+	return ring.Config{
+		Name:              "ruler", // Ensures separate ring metrics for ruler ring
+		Consul:            consulCfg,
+		Store:             cfg.Store,
+		HeartbeatTimeout:  cfg.HeartbeatTimeout,
+		ReplicationFactor: 1, // Ruler will never use a RF > 1
+		Mock:              cfg.Mock,
+	}
```
Comment (tomwilkie) [inline pkg/ruler/ring.go]: Lets have the ring's registerflags func take a prefix so we can have multiple of them, and then lets use an independant one for the ruler.  What we've got here won't give us the operation flexibility.

File: pkg/ruler/ruler.go [hunk 9]
```diff
@@ -116,92 +120,43 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
...
-func NewRuler(cfg Config, engine *promql.Engine, queryable storage.Queryable, d *distributor.Distributor) (*Ruler, error) {
+func NewRuler(cfg Config, engine *promql.Engine, queryable storage.Queryable, d *distributor.Distributor, rulesAPI RuleStore) (*Ruler, error) {
 	ncfg, err := buildNotifierConfig(&cfg)
 	if err != nil {
 		return nil, err
 	}
-	return &Ruler{
+
+	ruler := &Ruler{
+		cfg:           cfg,
```
Comment (tomwilkie) [inline pkg/ruler/ruler.go]: Now you're embedding the config (good), can you remove the fields that are just copied out of config?

File: pkg/ruler/ruler.go [hunk 10]
```diff
@@ -210,71 +165,63 @@ func NewRuler(cfg Config, engine *promql.Engine, queryable storage.Queryable, d
...
+			return nil, err
 		}
 	}
-	amConfig := &config.AlertmanagerConfig{
-		Scheme:                 u.Scheme,
-		PathPrefix:             u.Path,
-		Timeout:                model.Duration(rulerConfig.NotificationTimeout),
-		ServiceDiscoveryConfig: sdConfig,
+
+	if cfg.NumWorkers <= 0 {
```
Comment (tomwilkie) [inline pkg/ruler/ruler.go]: Can you put this kind of validation first pls?

File: pkg/ruler/ruler.go [hunk 11]
```diff
@@ -210,71 +165,63 @@ func NewRuler(cfg Config, engine *promql.Engine, queryable storage.Queryable, d
...
+	r.notifiersMtx.Unlock()
 
-	return promConfig, nil
+	r.scheduler.Stop()
+	for _, w := range r.workers {
+		w.Stop()
+	}
+
+	r.lifecycler.Shutdown()
+	r.ring.Stop()
```
Comment (tomwilkie) [inline pkg/ruler/ruler.go]: These will be nil if sharding isn't enabled, right?

File: pkg/ruler/ruler.go [hunk 12]
```diff
@@ -365,6 +312,13 @@ func (r *Ruler) getOrCreateNotifier(userID string) (*notifier.Manager, error) {
 func (r *Ruler) Evaluate(userID string, item *workItem) {
 	ctx := user.InjectOrgID(context.Background(), userID)
 	logger := util.WithContext(ctx, util.Logger)
+	if r.cfg.EnableSharding {
+		owned := r.checkRule(item.hash)
+		if !owned {
```
Comment (tomwilkie) [inline pkg/ruler/ruler.go]: You can compress this into one if.  I don't like nested ifs:

```go
if r.cfg.EnableSharding && !r.checkRule(item.hash) {
...
```

File: pkg/ruler/ruler.go [hunk 13]
```diff
@@ -384,109 +338,38 @@ func (r *Ruler) Evaluate(userID string, item *workItem) {
...
-		go w.Run()
-	}
-	level.Info(util.Logger).Log("msg", "ruler up and running")
-}
-
-// Stop the server.
-func (s *Server) Stop() {
-	for _, w := range s.workers {
-		w.Stop()
+func (r *Ruler) checkRule(hash uint32) bool {
```
Comment (tomwilkie) [inline pkg/ruler/ruler.go]: Might want to call this "ownsRule"?

File: pkg/ruler/ruler.go [hunk 14]
```diff
@@ -384,109 +338,38 @@ func (r *Ruler) Evaluate(userID string, item *workItem) {
...
+	rlrs, err := r.ring.Get(hash, ring.Read)
+	// If an error occurs keep evaluating a rule as if it is owned
+	// better to have extra datapoints for a rule than none at all
+	if err != nil {
+		ringCheckErrors.Inc()
```
Comment (tomwilkie) [inline pkg/ruler/ruler.go]: Can you log here too pls?

File: pkg/ruler/scheduler.go [hunk 15]
```diff
@@ -90,7 +96,7 @@ type userConfig struct {
 type groupFactory func(userID string, groupName string, rls []rules.Rule) (*group, error)
 
 type scheduler struct {
-	rulesAPI           RulesAPI
+	ruleStore          RuleStore
```
Comment (tomwilkie) [inline pkg/ruler/scheduler.go]: For simplicity, might be best not to rename this type?  Feel free to do it in another PR.

File: pkg/ruler/scheduler.go [hunk 16]
```diff
@@ -134,17 +139,17 @@ func (s *scheduler) Run() {
...
 			return
 		}
 	}
 }
 
 func (s *scheduler) Stop() {
-	close(s.stop)
+	close(s.done)
 	s.q.Close()
-	<-s.done
```
Comment (tomwilkie) [inline pkg/ruler/scheduler.go]: This double check is to give the scheduler time to shutdown.  Why remove it?

File: pkg/ruler/scheduler.go [hunk 17]
```diff
@@ -262,19 +268,29 @@ func (s *scheduler) addUserConfig(now time.Time, hasher hash.Hash64, generation
 				level.Warn(util.Logger).Log("msg", "scheduler: failed to create group for user", "user_id", userID, "group", group, "err", err)
 				return
 			}
-			workItems = append(workItems, workItem{userID, group, g, evalTime, generation})
+			ringHasher.Reset()
+			ringHasher.Write([]byte(userID + ":" + group))
```
Comment (tomwilkie) [inline pkg/ruler/scheduler.go]: I'd just do:
```
ringHasher.Write([]byte(userID))
ringHasher.Write([]byte(group))
``

The separate doesn't add anything here.

File: pkg/ruler/worker.go [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+	Run()
+	Stop()
+}
+
+type worker struct {
+	scheduler *scheduler
+	ruler     *Ruler
+
+	done       chan struct{}
+	terminated chan struct{}
```
Comment (tomwilkie) [inline pkg/ruler/worker.go]: `quit` & `done` are the names we tend to use.  lets be consistent.

**Comment by tomwilkie** [comment]
This looks really good!  Lots of small nits and we need to chat about how we handle config on Cortex.  When you get this green & all the points address, give me a ping and we can do a final pass.

File: pkg/ruler/lifecycle.go (hunk 7 above)
Comment (jtlisi) [reply pkg/ruler/lifecycle.go]: This basically just hangs the ruler til a new one joins the ring. A noop would probably work just as well.

File: pkg/ring/lifecycler.go (hunk 4 above)
Comment (jtlisi) [reply pkg/ring/lifecycler.go]: Addr is used by the ruler to verify ownership, I would rather use ID but that was not in the IngesterDesc.

File: pkg/ruler/scheduler.go (hunk 16 above)
Comment (jtlisi) [reply pkg/ruler/scheduler.go]: It's not needed since the scheduler is gracefully shut down later by the ruler during the flush stage of the lifecycle. It empties the scheduling queue as a noop currently and once the scheduling_queue closes it's next channel or a set timeout is hit, it will complete the flush and finish shutting down.

**Comment by jtlisi** [comment]
@tomwilkie I think this is ready for another look

File: pkg/ruler/worker.go [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+func (w *worker) Run() {
+	defer close(w.done)
+	for {
+		select {
+		case <-w.quit:
+			return
+		default:
+		}
+		waitStart := time.Now()
+		blockedWorkers.Inc()
```
Comment (gouthamve) [inline pkg/ruler/worker.go]: I see these metrics (`blockedWorkers`, `evalLatency`, `workerIdleTime`) are defined in `ruler.go`, can you move them here?

Comment (jtlisi) [reply pkg/ruler/worker.go]: 👍

**Comment by jtlisi** [comment]
I rebased this on top of master to work with a single binary. This should be ready for another look @tomwilkie

**Comment by tomwilkie** [comment]
LGTM! Thanks for you work on this Jacob.

**Comment by bboreham** [comment]
I have edited the description to mention some breaking changes to metrics.
